### ISSUE TITLE
CentOS 7 Installation Errors

### ISSUE THREAD

**Issue by npel21** [pyopencl/pyopencl#131]
I have been trying to get PyOpenCL version 2015.2.4 to install for a couple weeks now, but cannot seem to figure out the issue. I am attempting the install on a CentOS 7 server with GCC version 4.8.3 using Python 2.7.11 Anaconda 4.0 64 bit. I am using an Altera SDK for OpenCL version 16.0 and specifying the locations of the inc_dir and lib_dir when I run python configure.py. I have all the necessary packages installed.

The errors I am getting are:
```
in file included from .../src/c_wrapper/wrap.cl.h:88:16: error: redefinition of 'struct _cl_image_desc' typedef struct _cl_image_desc {

in file included from .../include/CL/cl_ext+altera.h:414:16: error: previous definition of 'struct _cl_image_desc' typedef struct _cl_image_desc {

in file included from .../src/c_wrapper/wrap_cl.h:99:16: error: invalid type in declaration before ';' token } cl_image_desc;

in file included from .../include/CL/cl_ext+altera.h:428:3: error: 'cl_image_desc' has a previous declaration as 'typedef struct _cl_image_desc cl_image_desc' } cl_image_desc;

error: command 'gcc' failed with exit status 1
make: *** [install] Error 1
```
I am not sure why these redefinition errors are happening and after hours of Googling I have not found much to help out. I am hoping someone could help me solve this issue.

Thanks.

**Comment by inducer** [comment]
Hum, this is odd. `cl_image_desc` is only supposed to be present in OpenCL 1.2 and later, as indicated by the `CL_VERSION_1_2` macro. Your ICD loader (may have come with the Altera SDK) apparently supplies CL headers that do have `cl_image_desc` but do not advertise support for CL 1.2. Could you share the `CL/cl.h` you are using (and what package it came from)?

**Comment by npel21** [comment]
I can't attach the file as my machine to access the internet at work is isolated. I commented out the image support code and PyOpenCL was able to install. I am running a hello world program and I keep getting errors from line 1 when I import pyopencl. First it was libopencl.so could not be found, so I updated my Path to include the CL lib folder. Now I am getting an error that clEnqueueReadBufferRect is an undefined symbol. I'm thinking PyOpenCL just does not play nice with Alter's OpenCL libraries so I guess I will be moving onto foreign function interface instead unfortunately.

**Comment by inducer** [comment]
One more thing you could try is to run

```
./configure.py --cl-pretend-version=1.1
rm -Rf build
python setup.py install
```

to rebuild PyOpenCL while targeting CL 1.1. Another option would be to use the [OCL ICD Loader](https://github.com/OCL-dev/ocl-icd) instead of the one you're currently using. That should just fix your issues without the need to muck with PyOpenCL.

**Comment by nachiket** [comment]
So I encountered a similar error with Xilinx flow, and realized both the FPGA vendors supply OpenCL implementations that are probably a bit too outdated OpenCL 1.0?

To circumvent the issue, I faked the include paths to a more compliant version -- NVIDIA's. The code then builds fine, and links against the Xilinx or Altera libraries.

I added the following line to siteconf.py to replace the one that points to Xilinx/Altera. DO NOT EDIT CL_LIB_DIR..

>     CL_INC_DIR = ['/opt/NVIDIA/cuda-7.5/include']

This should helpfully close the issue. And hope that the FPGA vendors release compliant OpenCL stacks in the future!

**Comment by inducer** [comment]
FWIW, I just fixed compilation against headers for CL 1.0, although I don't think that's what the problem is here--the above issue sounds more like a broken header shipped by Altera. I'll close this issue for now. Feel free to reopen if new information comes to light.
